Extension/src/Debugger/configurationProvider.ts

```diff
@@ -71,67 +80,149 @@ export class QuickPickConfigurationProvider implements vscode.DebugConfiguration
 
         const selection: MenuItem | undefined = await vscode.window.showQuickPick(items, {placeHolder: localize("select.configuration", "Select a configuration")});
         if (!selection) {
-            throw Error(localize("debug.configuration.selection.canceled", "Debug configuration selection canceled")); // User canceled it.
+            // throw Error(localize("debug.configuration.selection.canceled", "Debug configuration selection canceled")); // User canceled it.
+            Telemetry.logDebuggerEvent(DebuggerEvent.debugPanel, { "debugType": "debug", "folderMode": folder ? "folder" : "singleMode", "cancelled": "true" });
+            return []; // User canceled it.
         }
-        if (selection.label.startsWith("cl.exe")) {
-            if (!process.env.DevEnvDir) {
-                vscode.window.showErrorMessage(localize("cl.exe.not.available", "{0} build and debug is only usable when VS Code is run from the Developer Command Prompt for VS.", "cl.exe"));
-                return [selection.configuration];
-            }
-        }
-        if (selection.label.indexOf(buildAndDebugActiveFileStr()) !== -1 && selection.configuration.preLaunchTask) {
-            try {
-                await cppBuildTaskProvider.checkBuildTaskExists(selection.configuration.preLaunchTask);
-                if (selection.configuration.miDebuggerPath) {
-                    if (!await util.checkFileExists(selection.configuration.miDebuggerPath)) {
-                        vscode.window.showErrorMessage(localize("miDebuggerPath.not.available", "miDebuggerPath does not exist: {0}. Has a debugger been installed?", selection.configuration.miDebuggerPath));
-                        throw new Error();
-                    }
-                }
-                await vscode.debug.startDebugging(folder, selection.configuration);
-                Telemetry.logDebuggerEvent("buildAndDebug", { "success": "true" });
-            } catch (e) {
-                Telemetry.logDebuggerEvent("buildAndDebug", { "success": "false" });
-            }
+        if (!this.isClAvailable(selection.label)) {
+            return [selection.configuration];
         }
+
         return [selection.configuration];
     }
 
+    /**
+	 * Error checks the provided 'config' without any variables substituted.
+	 */
     resolveDebugConfiguration(folder: vscode.WorkspaceFolder | undefined, config: vscode.DebugConfiguration, token?: vscode.CancellationToken): vscode.ProviderResult<vscode.DebugConfiguration> {
-        return this.underlyingProvider.resolveDebugConfiguration ? this.underlyingProvider.resolveDebugConfiguration(folder, config, token) : undefined;
+        if (!config || !config.type) {
+            // When DebugConfigurationProviderTriggerKind is Dynamic, this function will be called with an empty config.
+            // Providing debug configs, and debugging should be called manually.
+            this.provideDebugConfigurations(folder).then(async configs => {
```

Comment:
This doesn't get returned, so the function returns `undefined` for this branch.

Reply:
it doesn't work if we return from here. This is the error:
![image](https://user-images.githubusercontent.com/16143955/154572075-4e11696e-e44f-4598-9eb2-2c58960d1054.png)

Reply:
Is that because we explicitly call `this.startDebugging` in the `else` case?